reconcile.go

```diff
@@ -234,54 +243,107 @@ func (a *Autopilot) pruneDeadServers() error {
 
 	state := a.GetState()
 
-	failed, voters, err := a.getFailedServers()
+	failed, vr, err := a.getFailedServers()
 	if err != nil || failed == nil {
 		return err
 	}
 
 	failed = a.promoter.FilterFailedServerRemovals(conf, state, failed)
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// remove stale non-voters
+	staleNonVoters := vr.FilterStale(failed.StaleNonVoters)
+	toRemove := a.adjudicateRemoval(staleNonVoters, vr.PotentialVoters)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.Remove(toRemove)
 
-	// remove stale non voters
-	for _, id := range failed.StaleNonVoters {
-		a.logger.Debug("removing stale raft server from configuration", "id", id)
-		if err := a.removeServer(id); err != nil {
-			return err
-		}
+	// Remove stale voters
+	staleVoters := vr.FilterStale(failed.StaleVoters)
+	toRemove = a.adjudicateRemoval(staleVoters, vr.PotentialVoters)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.Remove(toRemove)
 
-	maxRemoval := (voters - 1) / 2
+	// remove failed non-voters
+	failedNonVoters := vr.FilterFailed(failed.FailedNonVoters)
+	toRemove = a.adjudicateRemoval(failedNonVoters, vr.PotentialVoters)
+	a.removeFailedServers(failed.Get(toRemove, false))
+	vr.Remove(toRemove)
 
-	for _, id := range failed.StaleVoters {
-		if voters-1 < int(conf.MinQuorum) {
-			a.logger.Debug("will not remove server as it would leave less voters than the minimum number allowed", "id", id, "min", conf.MinQuorum)
-		} else if maxRemoval < 1 {
-			a.logger.Debug("will not remove server as removal of a majority of servers is not safe", "id", id)
-		} else if err := a.removeServer(id); err != nil {
-			return err
-		} else {
+	// remove failed voters
+	failedVoters := vr.FilterFailed(failed.FailedVoters)
+	toRemove = a.adjudicateRemoval(failedVoters, vr.PotentialVoters)
+	a.removeFailedServers(failed.Get(toRemove, true))
+	vr.Remove(toRemove)
+
+	return nil
+}
+
+func (a *Autopilot) adjudicateRemoval(s VoterRegistry, voterCountProvider func() int) []raft.ServerID {
```

Comment:
I find this approach a little confusing, whereby we take two arguments, one of which is a VoterRegistry, another is function pointer which in practice is a method to a different VoterRegistry.  I would again advocate for separating the notion of the voter registry as its own thing instead of letting it creep into all these places.  `s` here can just be a slice of serverIDs, and the second param can be used to look up whether a member of s is a potential voter and what the total voter count is.

Reply:
Updated to: `func (a *Autopilot) adjudicateRemoval(ids []raft.ServerID, vr *VoterRegistry) []raft.ServerID {`